### PR TITLE
Local adjustments - Denoise - improvments with Non Local means issue #6087

### PR DESCRIPTION
**Pull request by Desmis**
All comments are in issue

Jacques

### CODE REVIEW

File: rtengine/iplocallab.cc [hunk 1]
```diff
@@ -10872,6 +10872,16 @@ void ImProcFunctions::detail_mask(const array2D<float> &src, array2D<float> &mas
 // adapted to Rawtherapee Local adjustments J.Desmis january 2021
 //
 
+    static inline double igammalog    (double x, double p, double s, double g2, double g4)
+    {
+        return x <= g2 ? x / s : exp(log((x + g4) / (1.+ g4)) * p);//continuous
```
Comment (heckflosse) [inline rtengine/iplocallab.cc]: This smells really expensive. I will have a look

**Comment by Desmis** [comment]
@heckflosse 

OK, thank you

Jacques

File: rtengine/iplocallab.cc (hunk 1 above)
Comment (Thanatomanic) [reply rtengine/iplocallab.cc]: Mathematically this can also be `pow((x+g4)/(1.+g4),p)` right?

Comment (heckflosse) [reply rtengine/iplocallab.cc]: right, or better (performance-wise) pow_F instead of pow

**Comment by Desmis** [comment]
@heckflosse  @Thanatomanic 

I compile, works fine :)

Thank you very much

Jacques

**Comment by Desmis** [comment]
If nobody objects, I will merge this PR tomorrow :)

Jacques

**Comment by heckflosse** [comment]
@Desmis Jacques, please wait. I'm just trying further speedups for the (i)gammalog loops...

**Comment by Desmis** [comment]
No problem :)
Jacques

Le jeu. 4 févr. 2021 19:40, Ingo Weyrich <notifications@github.com> a
écrit :

**Comment by heckflosse** [comment]
@Desmis Jacques, please test ;-)

**Comment by heckflosse** [comment]
I forgot to mention my measures:

I measured using a 45.6 MP file

Original implementation of (i)gammalog loops was
~1250 ms per loop

My first change (using xexpf, xlogf, pow_f) reduced that to
~ 210 ms per loop

The SSE code further reduced that to

~ 80 ms per loop

**Comment by Desmis** [comment]
@heckflosse 
Hello Ingo

I just tested your improvment.
Compile fine

No differences with TIF
Same improvement (in percentage) as you for the acceleration of the processing on my 15Mb test file.
Excellent optimization work - as usual  :)

Thank you very much.

Do you have any other improvements to make?

Jacques

**Comment by heckflosse** [comment]
> Do you have any other improvements to make?

No. Ok to merge from me

**Comment by Desmis** [comment]
Ok thank you, I will merge now :)

Jacques

**Pull request merged**